src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/DictionaryEqualityComparer.cs

```diff
@@ -0,0 +1,118 @@
+        private DictionaryEqualityComparer()
+        {
+        }
+
+        /// <summary>
+        /// Gets a dictionary equality comparer instance appropriate for dictionaries that use the default key comparer for the <typeparamref name="TKey"/> type.
+        /// </summary>
+        internal static IEqualityComparer<IImmutableDictionary<TKey, TValue>> Instance
+        {
+            get { return defaultInstance; }
+        }
+
+        /// <summary>
+        /// Checks two dictionaries for equality.
+        /// </summary>
+        public bool Equals(IImmutableDictionary<TKey, TValue> x, IImmutableDictionary<TKey, TValue> y)
+        {
+            return AreEquivalent(x, y);
+        }
+
+        /// <summary>
+        /// Calculates a hash code for a dictionary.
+        /// </summary>
+        public int GetHashCode(IImmutableDictionary<TKey, TValue> obj)
+        {
+            int hashCode = 0;
+
+            var concreteDictionary1 = obj as ImmutableDictionary<TKey, TValue>;
+            IEqualityComparer<TKey> keyComparer = concreteDictionary1 != null ? concreteDictionary1.KeyComparer : EqualityComparer<TKey>.Default;
+            IEqualityComparer<TValue> valueComparer = concreteDictionary1 != null ? concreteDictionary1.ValueComparer : EqualityComparer<TValue>.Default;
+            if (obj != null)
+            {
+                foreach (var pair in obj)
+                {
+                    hashCode += keyComparer.GetHashCode(pair.Key) + valueComparer.GetHashCode(pair.Value);
+                }
+            }
+
+            return hashCode;
+        }
+        /// <summary>
+        /// Tests two dictionaries to see if their contents are identical.
+        /// </summary>
+        private static bool AreEquivalent(IImmutableDictionary<TKey, TValue> dictionary1, IImmutableDictionary<TKey, TValue> dictionary2)
+        {
+            Requires.NotNull(dictionary1, "dictionary1");
+            if (dictionary1 == dictionary2)
+            {
+                return true;
+            }
+
+            var concreteDictionary1 = dictionary1 as ImmutableDictionary<TKey, TValue>;
+            IEqualityComparer<TValue> valueComparer = concreteDictionary1 != null ? concreteDictionary1.ValueComparer : EqualityComparer<TValue>.Default;
+            return AreEquivalent((IReadOnlyDictionary<TKey, TValue>)dictionary1, (IReadOnlyDictionary<TKey, TValue>)dictionary2, valueComparer);
+        }
+
+        /// <summary>
+        /// Tests two dictionaries to see if their contents are identical.
+        /// </summary>
+        private static bool AreEquivalent(IReadOnlyDictionary<TKey, TValue> dictionary1, IReadOnlyDictionary<TKey, TValue> dictionary2, IEqualityComparer<TValue> valueComparer)
+        {
+            Requires.NotNull(valueComparer, "valueComparer");
+
+            if (dictionary1 == dictionary2)
+            {
+                return true;
+            }
+
+            if ((dictionary1 == null) ^ (dictionary2 == null)) // XOR
```

Comment:
Also, the only way that this could return false is `dictionary2` is false. You're already checking `dictionary1` in the first `AreEquivalent`.